google-service.js

```diff
@@ -0,0 +1,96 @@
+const { google } = require("googleapis");
+const _ = require("lodash");
+const { GOOGLE_API_CLIENT_NAMES } = require("./consts");
+
+/**
+ * This function injects given Google API Clients with JWT Auth Client
+ * into the parameters of the passed function with created
+ */
+function injectGoogleApiClients(callback, apiClientNames) {
+  validateApiClientNames(apiClientNames);
+  const scopes = determineAuthScopes(apiClientNames);
+  return async (params) => {
+    validateCredentials(params.credentials);
+    const authClient = await createGoogleServiceAccountAuthClient(params.credentials, scopes);
+    const googleClients = createGoogleApiClients(apiClientNames, authClient);
+    return callback(googleClients, _.omit(params, "credentials"));
+  };
+}
+
+/**
+ * Each API Client requires different authorization scope
+ */
+function determineAuthScopes(apiClientNames) {
+  const apiClientsRequiredScopesMap = new Map([
+    [GOOGLE_API_CLIENT_NAMES.SHEETS, ["https://www.googleapis.com/auth/spreadsheets"]],
+    [GOOGLE_API_CLIENT_NAMES.DRIVE, ["https://www.googleapis.com/auth/drive"]],
+  ]);
+  return apiClientNames.map((apiClientName) => (
+    apiClientsRequiredScopesMap.get(apiClientName)
+  )).flat();
+}
+
+function validateCredentials(credentialsObject) {
+  if (!credentialsObject) {
+    throw new Error("Google Service Account credentials are required. Please specify them in the action's parameters or plugin's settings.");
+  }
+  // These properties are required for JWT Auth Client
+  const requiredProperties = ["client_email", "private_key"];
+  requiredProperties.forEach((prop) => {
+    if (!Reflect.has(credentialsObject, prop)) {
+      throw new Error(`Missing property "${prop}" in the Google Service Account credentials object.`);
+    }
+  });
+}
+
+async function createGoogleServiceAccountAuthClient(credentials, scopes) {
+  const jwtAuthClient = new google.auth.JWT(
+    credentials.client_email,
+    null,
+    credentials.private_key,
+    scopes,
+  );
+  await jwtAuthClient.authorize();
+  return jwtAuthClient;
+}
+
+function validateApiClientNames(apiClientNames) {
```

Comment:
This function is used only by `injectGoogleApiClients`, where it's the first function called and this file exports `injectGoogleApiClients` only. It would be easier to understand the whole processing flow if this function was placed right below `injectGoogleApiClients`.